msprime/demography.py

```diff
@@ -1119,3 +1119,144 @@ def num_epochs(self):
+            Return the oldest node descended from ``node`` in ``node_attrs``.
+            """
+            epoch_j, pop_k = node
+            a = epoch_j - 1
+            while a >= 0:
+                descendent = (a, pop_k)
+                if descendent in node_attrs:
+                    break
+                a -= 1
+            if a < 0:
+                descendent = None
+            return descendent
+
+        if leaves is None:
+            leaves = range(self.num_populations)
+        else:
+            possible_leaves = set(range(self.num_populations))
+            for leaf in leaves:
+                if leaf not in possible_leaves:
+                    raise ValueError(f"Leaf {leaf} is not a valid population ID.")
+        # Population IDs that have been incorporated into the graph.
+        interned = set(leaves)
+
+        graph = collections.defaultdict(dict)
+        node_attrs = dict()
+        merged_pops = set()
+        for j, epoch in enumerate(self.epochs):
+            for de in epoch.demographic_events:
+                if isinstance(de, MassMigration):
+                    source = get_descendent(node_attrs, (j, de.source))
+                    if source is None:
+                        assert de.source not in interned
+                        raise ValueError(
+                            f"{de.source} not in graph at epoch {j}. "
+                            f"Should {de.source} be a leaf node?"
+                        )
+                    if de.proportion == 1:
+                        merged_pops.add(de.source)
+                        interned.add(de.dest)
+                        dest = (j, de.dest)
+                        graph[dest][source] = {}
+                        end_time = max(node_attrs[source]["end_time"], de.time)
+                        node_attrs[source].update(end_time=end_time)
+                    else:
+                        dest = get_descendent(node_attrs, (j, de.dest))
+                        assert dest is not None
+                        node_attrs[dest]["pulse_in"][(source, de.time)] = de.proportion
+                        node_attrs[source]["pulse_out"][(dest, de.time)] = de.proportion
+
+                elif isinstance(de, (SimpleBottleneck, InstantaneousBottleneck)):
+                    raise ValueError(f"Unsupported event: {de.__class__.__name__}")
+            for k, pop in enumerate(epoch.populations):
+                if k in interned and k not in merged_pops:
+                    node = (j, k)
+                    assert node not in node_attrs
+                    node_attr = dict(
+                        start_size=pop.start_size,
+                        end_size=pop.end_size,
+                        growth_rate=pop.growth_rate,
+                        start_time=epoch.start_time,
+                        end_time=epoch.end_time,
+                        M_in=epoch.migration_matrix[k, :],
+                        M_out=epoch.migration_matrix[:, k],
```

Comment:
I *think* these are in and out wrt forwards time.

Reply:
I was wrong, it appears this correpsonds to in/out with a backwards-time convention. This seems appropriate.